### DISCUSSION TITLE
Heap size larger than resident size?

### DISCUSSION THREAD

**Discussion by andronat** [bloomberg/memray#246] [Q&A] [answered]
Hello everyone! I'm amazed with memray! It's an epic tool that I've been using extensively for the last month to benchmark some pytorch models!

I recently stumbled upon the following graph:

<img width="1109" alt="image" src="https://user-images.githubusercontent.com/793408/199138757-e38da664-abad-4e4d-b0b7-2219345b9357.png">

I'm not sure how can I interpret the above result. My initial idea was that probably part of the heap has been swapped by the OS, but my RAM is significantly larger than the sizes requested by python. Any pointers on how to understand the results?

**Comment by pablogsal** [comment, marked as answer]
Thanks for opening this discussion!

Without looking at the code is difficult to tell for sure but this is kind of behaviour normally indicates that your application is requesting a lot of memory that’s not using. When you call malloc() you increase your heap size because you retested the memory but until you actually touch touch that memory by writing something to it, the OS won’t actually give do anything important so your resident size won’t increase.

You can do an easy experiment to see this: with numpy, allocate a big array with numpy.empty() and another one with numpy.zeros(). In the first case your resident size won’t increase but in the second it will (because writing zeros to the array needs the actual memory).

Another possibility is heavy fragmentation of the allocator or preallocation of data structures but that’s more difficult to confirm specially without the original code.
